### ISSUE TITLE
Does VALI support GetMotionVectors method?

### ISSUE THREAD

**Issue by cheptil** [RomanArzumanyan/VALI#13]
Hello

In VPF there is GetMotionVectors method, does VALI support this method?

**Comment by RomanArzumanyan** [comment]
@cheptil 

Pleae check out `issue-13_get_mv` branch, it has the fix for your issue.
Here's the API usage example:
https://github.com/RomanArzumanyan/VALI/blob/9eb701aba840e979477e6dc59991f22923deca4f/tests/test_PyFFmpegDecoder.py#L171-L192

**Comment by cheptil** [comment]
Hello

Thank you very much for the update!

I have tested it with test_res_change.h264 example from tests and see that num_elems for second frame was not zero, but the mv array is empty

Is it some bug or I need get access to mv info in another way?

**Comment by RomanArzumanyan** [comment]
Hm, that’s interesting.

I don’t observe such behaviour on basic test video, let me check the resolution change.

**Comment by RomanArzumanyan** [comment]
@cheptil 

After bashing against custom numpy array types for a while I've switched `PyFfmpegDecoder.GetMotionVectors` method return data type to `List[nvc.MotionVector]`.

Looks like it fixes the problem, I've tested both basic and resolution change videos.

**Comment by RomanArzumanyan** [comment]
@cheptil 

This method uses FFMpeg API for side data retrieval. So it supports everything FFMpeg does.

**Comment by cheptil** [comment]
@RomanArzumanyan

I have tested new change and confirm that now GetMotionVectors method works!

Also I noticed that for hevc video this method return empty mv array

As far as I understand, VPF/VALI do not use Video Codec SDK, so there is no AppEncME mode, only FFMpeg API — am I right?

**Comment by RomanArzumanyan** [comment]
Hi @cheptil 

> I have tested new change and confirm that now GetMotionVectors method works!

Good news, thanks.

> Also I noticed that for hevc video this method return empty mv array

If VALI returns empty MV array for HEVC P or B frame please provide me with repro video, I'll check on my side.
Motion Vectors API retrieval is codec-agnostic so it shall work for H.264 / HEVC etc.

> As far as I understand, VPF/VALI do not use Video Codec SDK, so there is no AppEncME mode, only FFMpeg API — am I right?

VALI does rely on Video Codec SDK for HW-accelerated video enc / dec.
It also uses FFMpeg for demuxing video containers and providing CPU-accelerated decoder fallback. It may be useful when input isn't supported by GPU decoder or simply if you want to utilize all the HW you have.

**Comment by RomanArzumanyan** [comment]
Hi @cheptil 

After inspecting the code in debugger I found that FFMpeg doesn't extract motion vectors for your H.265 video, hence VALI returns you an empty array.

Most probably I'm mistaken in my assumption that FFMpeg allows to extract MV for variety of codecs. Looks like it only supports H.264. I've found similar questions on SO and ffmpeg mailing list:

- https://stackoverflow.com/questions/47318705/extract-motion-vectors-from-x265-hevc-encoded-video-with-ffmpeg-libavcodec
- https://ffmpeg.org/pipermail/libav-user/2015-June/008242.html

Also there's a MV extraction patch for HEVC which to my understanding wasn't merged to FFMpeg codebase:

- https://patchwork.ffmpeg.org/project/ffmpeg/patch/20200115091908.81946-2-kaveasaf@gmail.com/#50356

Long story short: I'm afraid `PyFFmpegDecoder` only allows you to get MV for H.264 videos as that's what FFMpeg supports under the hood.

**Comment by RomanArzumanyan** [comment]
@cheptil 

I need to merge linked PR, it will close this issue.
Please feel free to re-open.
